### PR TITLE
Add non-recursive collision algorithm

### PR DESCRIPTION
**Pull request by IgorBaratta**
Also remove some expensive xtensor computations, and avoid allocating intermediate arrays.

```python
import dolfinx.mesh
import ufl

from dolfinx import geometry
from dolfinx import common
from dolfinx import fem
from mpi4py import MPI

comm = MPI.COMM_WORLD
mesh = dolfinx.mesh.create_unit_cube(comm, 50, 50, 50)

el = ufl.FiniteElement("Lagrange", mesh.ufl_cell(), 2)
V = fem.FunctionSpace(mesh, el)

x = V.tabulate_dof_coordinates()

tree = geometry.BoundingBoxTree(mesh, mesh.geometry.dim)

with common.Timer("compute_collisions") as t:
    cell_candidates = geometry.compute_collisions(tree, x)

common.list_timings(comm, [common.TimingType.wall])
```

This branch:
``` 
compute_collisions                                                          |     1  1.582766  1.582766
```

main:
```
compute_collisions                                                          |     1  14.208511  14.208511
```

### CODE REVIEW

File: cpp/dolfinx/geometry/utils.cpp [hunk 1]
```diff
@@ -110,32 +125,44 @@ _compute_closest_entity(const geometry::BoundingBoxTree& tree,
 // Compute collisions with point (recursive)
 void _compute_collisions_point(
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:126]: Maybe add in/out params with description here

File: cpp/dolfinx/geometry/utils.cpp [hunk 2]
```diff
@@ -110,32 +125,44 @@ _compute_closest_entity(const geometry::BoundingBoxTree& tree,
...
+    else
+    {
+      // Check whether the point collides with child nodes (left and right)
+      bool left = point_in_bbox(tree.get_bbox(bbox[0]), p);
+      bool right = point_in_bbox(tree.get_bbox(bbox[1]), p);
+      if (left && right)
+      {
+        // If the point collides with the two child nodes, add a node to the
+        // stack (for later visiting) and continue the tree traversal with
+        // the left subtree
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:151]:
```suggestion
        // If the point collides with the both child nodes, add the right node to the
        // stack (for later visiting) and continue the tree traversal with
        // the left subtree
```

File: cpp/dolfinx/geometry/utils.cpp [hunk 3]
```diff
@@ -107,35 +123,53 @@ _compute_closest_entity(const geometry::BoundingBoxTree& tree,
...
-    entities.push_back(entity_index);
-  }
-  else
-  {
-    // Check both children
-    _compute_collisions_point(tree, p, bbox[0], entities);
-    _compute_collisions_point(tree, p, bbox[1], entities);
+    if (is_leaf(bbox))
+      // If box is a leaf node then add it to the list of colliding entities
+      entities.push_back(bbox[1]);
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:145]:
```suggestion
    if (is_leaf(bbox))
    {
      // If box is a leaf node then add it to the list of colliding entities
      entities.push_back(bbox[1]);
    }
```

File: cpp/dolfinx/geometry/utils.cpp [hunk 4]
```diff
@@ -107,35 +123,53 @@ _compute_closest_entity(const geometry::BoundingBoxTree& tree,
...
+      {
+        // If the point collides with both child nodes, add the right node to
+        // the stack (for later visiting) and continue the tree traversal with
+        // the left subtree
+        stack.push_back(bbox[1]);
+        next = bbox[0];
+      }
+      else if (left)
+        // Traverse the current node's left subtree
+        next = bbox[0];
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:161]:
```suggestion
      {
        // Traverse the current node's left subtree
        next = bbox[0];
      }
```

File: cpp/dolfinx/geometry/utils.cpp [hunk 5]
```diff
@@ -107,35 +123,53 @@ _compute_closest_entity(const geometry::BoundingBoxTree& tree,
...
+      else if (right)
+        // Traverse the current node's right subtree
+        next = bbox[1];
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:164]:
```suggestion
      else if (right)
      {
        // Traverse the current node's right subtree
        next = bbox[1];
      }
```

**Comment by IgorBaratta** [comment]
If this PR goes through, another one can follow doing the same thing with `_compute_collisions_tree`.

File: cpp/dolfinx/geometry/utils.cpp [hunk 6]
```diff
@@ -32,18 +33,33 @@ bool point_in_bbox(const xt::xtensor_fixed<double, xt::xshape<2, 3>>& b,
...
   constexpr double rtol = 1e-14;
-  auto eps = rtol * (xt::row(b, 1) - xt::row(b, 0));
-  return xt::all(x >= xt::row(b, 0) - eps)
-         and xt::all(x <= xt::row(b, 1) + eps);
+  double eps;
+  bool in = true;
+  for (int i = 0; i < 3; i++)
+  {
+    eps = rtol * (b(1, i) - b(0, i));
+    in &= x[i] >= (b(0, i) - eps);
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.cpp:41]: Maybe add a brief explanation of what's happening?

File: cpp/dolfinx/geometry/utils.cpp [hunk 7]
```diff
@@ -107,35 +123,59 @@ _compute_closest_entity(const geometry::BoundingBoxTree& tree,
...
+      }
+      else if (right)
+      {
+        // Traverse the current node's right subtree
+        next = bbox[1];
+      }
+    }
+    // If tree traversal reaches a dead end (box is a leaf node or no collision
+    // detected), check the stack for deferred subtrees.
+    if (next == -1 && !stack.empty())
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.cpp:174]: We generally use `and` rather than `&&`.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Should we change some `xt::xtensor_fixed<double, xt::xshape<3>>` to `std::array<double, x>` if we're making less use of xtension functions?

File: cpp/dolfinx/geometry/utils.cpp (hunk 7 above)
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:174]:
```suggestion
    if (next == -1 and !stack.empty())
```

File: cpp/dolfinx/geometry/utils.cpp [hunk 8]
```diff
@@ -29,6 +29,9 @@ constexpr bool is_leaf(const std::array<int, 2>& bbox)
   return bbox[0] == bbox[1];
 }
 //-----------------------------------------------------------------------------
+// A point `x` is inside a bounding box `b` if each component of its coordinates
+// lies within the range `[b(0,i), b(1,i)]` that defines the bounds of the
+// bounding box, b(0,i) <= x[i] <= b(1,i) for i = 0, 1, 2
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:34]:
```suggestion
/// A point `x` is inside a bounding box `b` if each component of its coordinates
/// lies within the range `[b(0,i), b(1,i)]` that defines the bounds of the
/// bounding box, b(0,i) <= x[i] <= b(1,i) for i = 0, 1, 2
```

File: cpp/dolfinx/geometry/utils.cpp [hunk 9]
```diff
@@ -45,6 +48,9 @@ bool point_in_bbox(const xt::xtensor_fixed<double, xt::xshape<2, 3>>& b,
   return in;
 }
 //-----------------------------------------------------------------------------
+// A bounding box "a" is contained inside another bounding box "b", if each
+// of its intervals [a(0,i), a(1,i)] is contained in [b(0,i), b(1,i)],
+// a(0,i) <= b(1, i) and a(1,i) >= b(0, i)
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.cpp:53]:
```suggestion
/// A bounding box "a" is contained inside another bounding box "b", if each
/// of its intervals [a(0,i), a(1,i)] is contained in [b(0,i), b(1,i)],
/// a(0,i) <= b(1, i) and a(1,i) >= b(0, i)
```

**Review by garth-wells — APPROVED** [review]

**Pull request merged**